File: frontend/api/tags.ts
import { apiClient } from "./client";
import type { Tag } from "./types";

export async function listTags(): Promise<Tag[]> {
  const response = await apiClient.get<Tag[]>("/api/tags");
  return response.data;
}

export async function createTag(name: string, color?: string | null, textColor?: string | null): Promise<Tag> {
  const response = await apiClient.post<Tag>("/api/tags", { name, color, textColor });
  return response.data;
}

export async function updateTag(
  tagId: number,
  name: string,
  color?: string | null,
  textColor?: string | null
): Promise<Tag> {
  const response = await apiClient.patch<Tag>(`/api/tags/${tagId}`, { name, color, textColor });
  return response.data;
}

export async function deleteTag(tagId: number): Promise<void> {
  await apiClient.delete(`/api/tags/${tagId}`);
}

export async function attachTag(gameId: number, tagId: number): Promise<Tag[]> {
  const response = await apiClient.post<Tag[]>(`/api/games/${gameId}/tags/${tagId}`);
  return response.data;
}

export async function detachTag(gameId: number, tagId: number): Promise<Tag[]> {
  const response = await apiClient.delete<Tag[]>(`/api/games/${gameId}/tags/${tagId}`);
  return response.data;
}

export async function attachTagToDevice(deviceId: number, tagId: number): Promise<Tag[]> {
  const response = await apiClient.post<Tag[]>(`/api/devices/${deviceId}/tags/${tagId}`);
  return response.data;
}

export async function detachTagFromDevice(deviceId: number, tagId: number): Promise<Tag[]> {
  const response = await apiClient.delete<Tag[]>(`/api/devices/${deviceId}/tags/${tagId}`);
  return response.data;
}

export async function attachTagToAccessory(accessoryId: number, tagId: number): Promise<Tag[]> {
  const response = await apiClient.post<Tag[]>(`/api/accessories/${accessoryId}/tags/${tagId}`);
  return response.data;
}

export async function detachTagFromAccessory(accessoryId: number, tagId: number): Promise<Tag[]> {
  const response = await apiClient.delete<Tag[]>(`/api/accessories/${accessoryId}/tags/${tagId}`);
  return response.data;
}

File: backend/app/api/routes/tags.py
from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_user, get_db
from app.schemas.tag import TagCreateRequest, TagResponse, tag_from_orm
from app.services import tag_service

router = APIRouter(tags=["tags"], dependencies=[Depends(get_current_user)])


@router.get("/api/tags", response_model=list[TagResponse])
def list_tags(db: Session = Depends(get_db)) -> list[TagResponse]:
    return [tag_from_orm(tag) for tag in tag_service.list_tags(db)]


@router.post("/api/tags", response_model=TagResponse, status_code=status.HTTP_201_CREATED)
def create_tag(body: TagCreateRequest, db: Session = Depends(get_db)) -> TagResponse:
    tag = tag_service.create_tag(db, body.name, body.color)
    return tag_from_orm(tag)


@router.delete("/api/tags/{tag_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_tag(tag_id: int, db: Session = Depends(get_db)) -> None:
    tag_service.delete_tag(db, tag_id)


@router.post("/api/games/{game_id}/tags/{tag_id}", response_model=list[TagResponse])
def attach_tag(game_id: int, tag_id: int, db: Session = Depends(get_db)) -> list[TagResponse]:
    tags = tag_service.attach_tag(db, game_id, tag_id)
    return [tag_from_orm(tag) for tag in tags]


@router.delete("/api/games/{game_id}/tags/{tag_id}", response_model=list[TagResponse])
def detach_tag(game_id: int, tag_id: int, db: Session = Depends(get_db)) -> list[TagResponse]:
    tags = tag_service.detach_tag(db, game_id, tag_id)
    return [tag_from_orm(tag) for tag in tags]


@router.post("/api/devices/{device_id}/tags/{tag_id}", response_model=list[TagResponse])
def attach_tag_to_device(device_id: int, tag_id: int, db: Session = Depends(get_db)) -> list[TagResponse]:
    tags = tag_service.attach_tag_to_device(db, device_id, tag_id)
    return [tag_from_orm(tag) for tag in tags]


@router.delete("/api/devices/{device_id}/tags/{tag_id}", response_model=list[TagResponse])
def detach_tag_from_device(device_id: int, tag_id: int, db: Session = Depends(get_db)) -> list[TagResponse]:
    tags = tag_service.detach_tag_from_device(db, device_id, tag_id)
    return [tag_from_orm(tag) for tag in tags]


@router.post("/api/accessories/{accessory_id}/tags/{tag_id}", response_model=list[TagResponse])
def attach_tag_to_accessory(accessory_id: int, tag_id: int, db: Session = Depends(get_db)) -> list[TagResponse]:
    tags = tag_service.attach_tag_to_accessory(db, accessory_id, tag_id)
    return [tag_from_orm(tag) for tag in tags]


@router.delete("/api/accessories/{accessory_id}/tags/{tag_id}", response_model=list[TagResponse])
def detach_tag_from_accessory(accessory_id: int, tag_id: int, db: Session = Depends(get_db)) -> list[TagResponse]:
    tags = tag_service.detach_tag_from_accessory(db, accessory_id, tag_id)
    return [tag_from_orm(tag) for tag in tags]
